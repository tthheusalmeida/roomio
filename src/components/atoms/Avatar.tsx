import { mergeClassNames } from "@/utils/classNames";
import Image from "next/image";

interface AvatarProps {
  src: string;
  alt?: string;
  name: string;
  size?: "small" | "medium" | "large";
  className?: string;
}

interface SizeInterface {
  width: number;
  height: number;
}

const SIZES: Record<string, SizeInterface> = {
  small: {
    width: 28,
    height: 28,
  },
  medium: {
    width: 52,
    height: 52,
  },
  large: {
    width: 120,
    height: 120,
  },
};

export default function Avatar({
  src,
  alt,
  name,
  size = "small",
  className,
}: AvatarProps) {
  const { width, height } = SIZES[size];
  const description = alt ? alt : `${name} avatar`;

  return (
    <Image
      src={src as string}
      alt={description}
      width={width}
      height={height}
      className={mergeClassNames("rounded-full relative self-start", className)}
    />
  );
}
