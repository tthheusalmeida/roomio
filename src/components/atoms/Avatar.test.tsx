import { render, screen } from "@testing-library/react";
import Avatar from "./Avatar";

jest.mock("next/image", () => (props: any) => {
  return <img {...props} />;
});

describe("Avatar component", () => {
  it("renders with default size (small) and automatic alt", () => {
    const { asFragment } = render(<Avatar src="/avatar.png" name="Matheus" />);

    const img = screen.getByRole("img");
    expect(img).toBeInTheDocument();
    expect(img).toHaveAttribute("src", "/avatar.png");
    expect(img).toHaveAttribute("alt", "Matheus avatar");
    expect(img).toHaveAttribute("width", "28");
    expect(img).toHaveAttribute("height", "28");

    expect(asFragment()).toMatchSnapshot();
  });

  it("uses custom alt when provided", () => {
    const { asFragment } = render(
      <Avatar src="/avatar.png" name="Matheus" alt="Profile picture" />
    );
    const img = screen.getByRole("img");
    expect(img).toHaveAttribute("alt", "Profile picture");

    expect(asFragment()).toMatchSnapshot();
  });

  it("renders in medium size", () => {
    const { asFragment } = render(
      <Avatar src="/avatar.png" name="Matheus" size="medium" />
    );
    const img = screen.getByRole("img");
    expect(img).toHaveAttribute("width", "52");
    expect(img).toHaveAttribute("height", "52");

    expect(asFragment()).toMatchSnapshot();
  });

  it("applies extra classes passed via className", () => {
    const { asFragment } = render(
      <Avatar src="/avatar.png" name="Matheus" className="extra-class" />
    );
    const img = screen.getByRole("img");
    expect(img.className).toMatch(/extra-class/);

    expect(asFragment()).toMatchSnapshot();
  });
});
