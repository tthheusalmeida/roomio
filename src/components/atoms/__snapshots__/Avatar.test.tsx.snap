// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Avatar component applies extra classes passed via className 1`] = `
<DocumentFragment>
  <img
    alt="Matheus avatar"
    class="rounded-full relative self-start extra-class"
    height="28"
    src="/avatar.png"
    width="28"
  />
</DocumentFragment>
`;

exports[`Avatar component renders in medium size 1`] = `
<DocumentFragment>
  <img
    alt="Matheus avatar"
    class="rounded-full relative self-start"
    height="52"
    src="/avatar.png"
    width="52"
  />
</DocumentFragment>
`;

exports[`Avatar component renders with default size (small) and automatic alt 1`] = `
<DocumentFragment>
  <img
    alt="Matheus avatar"
    class="rounded-full relative self-start"
    height="28"
    src="/avatar.png"
    width="28"
  />
</DocumentFragment>
`;

exports[`Avatar component uses custom alt when provided 1`] = `
<DocumentFragment>
  <img
    alt="Profile picture"
    class="rounded-full relative self-start"
    height="28"
    src="/avatar.png"
    width="28"
  />
</DocumentFragment>
`;
